docs/design-documents/20221027-builtin-config-validation.md

```diff
@@ -0,0 +1,144 @@
+# Built-in config validation
+
+This design document discusses the best approach for implementing built-in config validations.
+
+## The problem
+
+Config validations are not enforced in Conduit or the SDK, and developers have the freedom of either validating the
+configuration for their connectors or not. So there’s a chance that some configurations could be marked as mandatory,
+but not enforced in any way.
+
+## What we currently do
+
+We have a `validate` endpoint for Conduit, which users can use to make sure that the connector configurations are valid,
+without actually creating the connector.
+
+**Cons**:
+
+- It’s just an option for the user to use this.
+- It calls the `Configure` function, which is written by the developer, so still not guaranteed to enforce the validations.
+
+We also have this proto message:
+
+```protobuf
+message PluginSpecifications {
+  message Parameter {
+    // Validation to be made on the parameter.
+    message Validation{
+      enum Type {
+        TYPE_UNSPECIFIED = 0;
+        // Parameter must be present.
+        TYPE_REQUIRED = 1;
+        // Parameter must be greater than {value}.
+        TYPE_GREATER_THAN = 2;
+        // Parameter must be less than {value}.
+        TYPE_LESS_THAN = 3;
+        // Parameter must be included in the comma separated list {value}.
+        TYPE_INCLUSION = 4;
+        // Parameter must not be included in the comma separated list {value}.
+        TYPE_EXCLUSION = 5;
+        // Parameter must match the regex {value}.
+        TYPE_REGEX = 6;
```

Comment:
Is this the list of validations we plan to support in the first step?

Reply:
yes, first step will allow the developer to add these validations to the params, second step will be the one validating